R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+        .increasing_in_interval = .data$plus_t2 > .data$t0,
+        .is_monotonic = .data$.increasing_in_interval & .data$.increasing_overall,
+        monotonic_factor = dplyr::if_else(.data$.is_monotonic, 1, -1),
+        .increasing_overall = NULL,
+        .increasing_in_interval = NULL,
+        .is_monotonic = NULL,
+        t0 = NULL,
+        plus_t1 = NULL,
+        plus_t2 = NULL,
+        end_year = NULL
+      ) %>%
+      ungroup()
+  }
+
+  left_join(data, monotonic_factors, by = groups)
+}
+
+add_tdm <- function(data, groups) {
+  data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    mutate(
+      .numerator = .data$scenario_production_plus_t2 - .data$portfolio_production_plus_t1,
+      .denominator = .data$scenario_production_plus_t2 - .data$scenario_production_t0,
+      # TODO: This case we
+      tdm_tech = ifelse(
+        .data$.denominator == 0,
+        0,
+        max(0, (.data$.numerator / .data$.denominator) * .data$monotonic_factor) * 2
+      ),
+      .numerator = NULL,
+      .denominator = NULL
+    ) %>%
+    select(.data$tdm_tech, all_of(groups)) %>%
+    distinct() %>%
+    ungroup()
+}
+
+add_aggregate_tdm <- function(data, t0, groups) {
+  data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    ungroup(.data$technology) %>%
+    mutate(
+      tdm_sec = .data$plan_carsten * sum(.data$tdm_tech) / sum(.data$plan_carsten),
+      reference_year = t0
+    ) %>%
+    ungroup()
+}
+
+pre_format_data <- function(data, t0, t1, t2, groups) {
+  data %>%
+    filter(.data$year %in% c(t0, t0 + t1, t0 + t2)) %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    add_time_step(t0, t1, t2) %>%
+    select(
+      !!!rlang::syms(groups),
+      .data$time_step,
+      .data$monotonic_factor,
+      scenario_production = .data$scen_alloc_wt_tech_prod,
+      portfolio_production = .data$plan_alloc_wt_tech_prod,
+    ) %>%
+    pivot_wider(
+      names_from = .data$time_step,
+      values_from = all_of(c("scenario_production", "portfolio_production"))
+    ) %>%
+    ungroup()
+}
+
+tdm_prototype <- function() {
+  tibble(
+    technology = character(0),
+    ald_sector = character(0),
+    tdm_tech = numeric(0),
+    tdm_sec = numeric(0),
+    reference_year = integer(0),
+  )
+}
+
+check_calculate_tdm <- function(data, t0, t1, t2, groups) {
+  check_crucial_names(data, crucial_columns())
+
+  check_crucial_years(data, t0, t1, t2, groups)
+
+  check_unique_by_year_and_groups(data, groups)
+
+  invisible(data)
+}
+
+check_unique_by_year_and_groups <- function(data, groups) {
+  req_unique_columns <- c(
+    "scen_alloc_wt_tech_prod",
+    "plan_alloc_wt_tech_prod",
+    "plan_carsten"
+  )
+
+  data <- data %>%
+    group_by(!!!rlang::syms(c("year", groups))) %>%
+    dplyr::summarize(rows_by_year_and_group = dplyr::n()) %>%
+    mutate(rows_are_unique = .data$rows_by_year_and_group == 1)
+
+  ok <- all(data$rows_are_unique)
+  if (!ok) {
+    rlang::abort(
+      "multiple_values_per_year",
+      message = glue(
```

Comment:
The argument `message` is quite clever. You can get a nicer output. See `message` at `abort()`.

``` r
rlang::abort(c("Title.", x = "Problem.", i = "Hint"))
#> Error: Title.
#> x Problem.
#> ℹ Hint
```

<sup>Created on 2021-11-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>